Support for more AiProvider then OPenAi

dieriba opened the pull request:
<!-- ELLIPSIS_HIDDEN -->



> [!IMPORTANT]
> Add support for multiple AI providers, including OpenAI and Anthropic, with updates to backend, frontend, and database schema.
> 
>   - **Backend**:
>     - Add support for Anthropic AI alongside OpenAI in `ai.rs`.
>     - Modify database schema in `migrations` to store AI provider information.
>     - Remove `openai.rs` and integrate its functionality into `ai.rs`.
>   - **Frontend**:
>     - Update components like `ScriptGen.svelte`, `StepGen.svelte`, and `FlowModuleSchemaMap.svelte` to support multiple AI providers.
>     - Add `AiProviderTypes` to manage AI provider selection.
>     - Modify `TestAiKey.svelte` to test keys for different AI providers.
>     - Update `workspace_settings/+page.svelte` to allow selection between OpenAI and Anthropic.
>   - **Misc**:
>     - Update `stores.ts` to track AI provider and resource existence.
>     - Add new utility functions and types in `lib.ts` to handle AI provider logic.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=windmill-labs%2Fwindmill&utm_source=github&utm_medium=referral)<sup> for 8bd97d2219952c0834c6493bdb4c8250f6341ddb. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

HugoCasa commented on backend/windmill-api/openapi-deref.yaml:
file is handled automatically, no need to commit those changes

HugoCasa commented on backend/windmill-api/openapi-deref.json:
file is handled automatically, no need to commit those changes

HugoCasa commented on backend/.gitignore:
```diff
@@ -2,7 +2,6 @@ target/
 .env
 oauth.json
 oauth2.json
-windmill-api/openapi-deref.yaml
```
leave it

dieriba commented:
> **CLA Assistant Lite bot:** <br/>Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you sign our [Contributor License Agreement](https://github.com/windmill-labs/windmill/blob/master/CLA.md) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.<br/>
>    - - -
>    I have read the CLA Document and I hereby sign the CLA
>    - - -
>    <sub>You can retrigger this bot by commenting **recheck** in this Pull Request</sub>

I have read the CLA Document and I hereby sign the CLA

dieriba commented:
I have read the CLA Document and I hereby sign the CLA

dieriba commented:
**recheck**

HugoCasa commented on frontend/src/routes/(root)/(logged)/+layout.svelte:
```diff
@@ -227,12 +227,16 @@
-		workspacedOpenai.init(workspace)
+		OpenAi.workspacedOpenai.init(workspace)
+		AnthropicAi.workspacedAnthropic.init(workspace)
+		MistralAi.workspacedMistral.init(workspace)
```
can you combine all inits into a single function that you export from copilot/lib?

dieriba replied:
Done!

HugoCasa commented on frontend/src/lib/components/copilot/lib.ts:
```diff
@@ -255,57 +299,148 @@ const PROMPTS_CONFIGS = {
+	return 'type' in message
+}
+*/
+function isChatCompletionChunk(
```
still to do

HugoCasa reviewed (commented):
Can you share screenshots of the UI changes you made? (create workspace, workspace settings, etc...)

HugoCasa reviewed (changes requested):
^

dieriba commented:
> Can you share screenshots of the UI changes you made? (create workspace, workspace settings, etc...)

Here is the workspace settings screenshot, I added a toggle group button to choose between all differents providers, internally it just update the resource_type 
![Screenshot 2024-11-10 at 16-49-15 Workspace Settings Windmill](https://github.com/user-attachments/assets/207e4ab3-192b-4084-996d-8b6c31a6fe77).
And here is a screen shot of create workspace I added the same toggle button group which also update internaly the resource_type, and also resize the enable completion button which was in size='xs'.
![Screenshot 2024-11-10 at 16-52-22 Windmill](https://github.com/user-attachments/assets/4fb2f1b1-bbd8-488d-baa2-389cdbea9be5)

dawidmachon commented:
I will kindly ask if there is a chance to make this more agnostic for self-hosted users? We're often deploying own models on our infrastructure, which has OpenAI API. For example: https://github.com/TabbyML/tabby

Is there any chance to get this feature more agnostic to also include own instances deployed somewhere else that standard endpoints?

HugoCasa commented on backend/windmill-api/src/ai.rs:
```diff
@@ -0,0 +1,506 @@
+
+    #[derive(Clone, Deserialize, Debug)]
+    pub struct AnthropicCache {
+        pub api_key: String,
```
use apiKey instead for anthropic (https://hub.windmill.dev/resource_types/190/anthropic)

dieriba replied:
Done!

dieriba commented on frontend/src/lib/components/Dev.svelte:
```diff
@@ -108,12 +108,15 @@
-			workspacedOpenai.init(workspace, token)
+			OpenAi.workspacedOpenai.init(workspace, token)
+			MistralAi.workspacedMistral.init(workspace, token)
+			AnthropicAi.workspacedAnthropic.init(workspace, token)
```
Done

dieriba commented on frontend/src/routes/(root)/(logged)/workspace_settings/+page.svelte:
```diff
@@ -976,27 +988,35 @@
+				<ToggleButton value="openai" label="OpenAi" />
+				<ToggleButton value="anthropic" label="Anthropic" />
+				<ToggleButton value="mistral" label="Mistral" />
+			</ToggleButtonGroup>
```
I thought the same about UX, however I let it that way so user can see which key is set. For now I added your suggestion so it does actually reset aiResourceInitialPath each time a new option is selected, tell me your though about it.

HugoCasa replied:
It shouldn't be done on start. There is a "selected" event on <ToggleButtonGroup> and you can reset aiResourceInitialPath when dispatched: on:selected={() => {aiResourceInitialPath = ''}
So it would only be reset when changing the kind of resource. What do you think?

dieriba replied:
You're totally right, it will provide better readability about what actually happens when one event is selected, I have made the fix.

HugoCasa commented:
#4692
